Add OGCGeometry::estimateMemorySize API

mbasmanova opened the pull request:
#156

The new OGCGeometry::estimateMemorySize method returns estimates amount of memory used by the object in bytes. Here are the values for some test objects:

```
OGCConcreteGeometryCollection: 426
OGCMultiPolygon: 422
OGCPoint: 80
OGCLineString: 338
OGCPolygon: 354
OGCMultiLineString: 422
OGCMultiPoint: 192
```

randallwhitman commented:
When proposing a change that adds a dependency, it would be good to mention so upfront, along with info on license and size.

randallwhitman commented:
Does `org.openjdk.jol` work equally well with JDK/JRE other than OpenJDK, such as Oracle or IBM JRE?

stolstov commented:
The license seems to be GPL 2.0 for this. https://mvnrepository.com/artifact/org.openjdk.jol/jol-core
Seems to be incompatible with Apache.

stolstov commented:
I think the only way we could have this in geometry is by providing approximate calculations based of the object layout. For example:
'''
class Geometry
{
  private static final int CLASS_SIZE = 32; //something not zero
  long  estimateMemorySize() { return CLASS_SIZE; }
} 

class Point extends Geometry
{
  private static final int CLASS_SIZE_EXTRA = 8; //does not include Geometry part
  @Override
  long  estimateMemorySize() 
  { 
    long sz = super.estimateMemorySize() + Point.CLASS_SIZE_EXTRA; 
    if (m_attributes != null)
      sz += m_attributes.length;

    return sz;
  }
}

randallwhitman commented:
One idea would be a test-only dependency on `jol` that would vet the sizes returned by `estimateMemorySize` as proposed by Sergey in previous comment.

mbasmanova commented:
@randallwhitman @stolstov Randall, Sergey, thank you for review. I updated the PR to make `jol` test-only dependency, hardcode instance sizes and test these values. Please, take another look.

martint commented:
The license for JOL is actually GPL with Classpath Exception. It's the same license as OpenJDK.

http://hg.openjdk.java.net/code-tools/jol/file/03064c057dc9/LICENSE
http://hg.openjdk.java.net/code-tools/jol/file/03064c057dc9/jol-core/src/main/java/org/openjdk/jol/info/ClassLayout.java

```
"CLASSPATH" EXCEPTION TO THE GPL

Certain source files distributed by Oracle America and/or its affiliates are
subject to the following clarification and special exception to the GPL, but
only where Oracle has expressly included in the particular source file's header
the words "Oracle designates this particular file as subject to the "Classpath"
exception as provided by Oracle in the LICENSE file that accompanied this code."

    Linking this library statically or dynamically with other modules is making
    a combined work based on this library.  Thus, the terms and conditions of
    the GNU General Public License cover the whole combination.

    As a special exception, the copyright holders of this library give you
    permission to link this library with independent modules to produce an
    executable, regardless of the license terms of these independent modules,
    and to copy and distribute the resulting executable under terms of your
    choice, provided that you also meet, for each linked independent module,
    the terms and conditions of the license of that module.  An independent
    module is a module which is not derived from or based on this library.  If
    you modify this library, you may extend this exception to your version of
    the library, but you are not obligated to do so.  If you do not wish to do
    so, delete this exception statement from your version.
```

stolstov commented:
@mbasmanova That looks very neat. I'll review this a little more. 

I am still uneasy of having to link to the GPL project as a part of the build process (although it has an exception, you only see the exception if you cared to read their license to the end)

stolstov commented on src/test/java/com/esri/core/geometry/TestEstimateMemorySize.java:
```diff
@@ -0,0 +1,108 @@
+import com.esri.core.geometry.ogc.OGCPoint;
+import com.esri.core.geometry.ogc.OGCPolygon;
+import org.junit.Test;
+import org.openjdk.jol.info.ClassLayout;
```
@mbasmanova Could you add a comment here for now for the reference
//ClassLayout is GPL with Classpath exception, see http://openjdk.java.net/legal/gplv2+ce.html

mbasmanova replied:
@stolstov Sure. Added the suggested comment.

stolstov commented on src/main/java/com/esri/core/geometry/Geometry.java:
```diff
@@ -150,6 +150,23 @@ public int value() {
+		return attributes != null ? sizeOfDoubleArray(attributes.length) : 0;
+	}
+
+	protected static long estimateMemorySize(VertexDescription description)
```
@mbasmanova The estimateMemorySize(VertexDescription description) method should be removed. The description instance is shared between all instances of geometry.

mbasmanova replied:
@stolstov Good to know. I'll remove it.

stolstov commented on src/main/java/com/esri/core/geometry/ogc/OGCPoint.java:
```diff
@@ -77,6 +79,13 @@ public String geometryType() {
+	public long estimateMemorySize()
+	{
+		return SIZE_OF_OGC_POINT + (point != null ? point.estimateMemorySize() : 0) +
+				(esriSR != null ? esriSR.estimateMemorySize() : 0);
```
In a good implementation, the spatial reference instance should be shared between all instances of geometry coming the same dataset, so usually the esriSR size would not contribute to OGC types, but we don't enforce that on geometry side right now.

mbasmanova replied:
@stolstov Sergey, would you recommend keeping this or removing and hoping esriSR is actually shared?

stolstov replied:
@mbasmanova Let's exclude it, and add a doc to OGCGeometry.estimateMemorySize, mentioning that the size of spatial reference is not included, because we assume that it is shared.

mbasmanova commented:
@stolstov Sergey, thank you for review. I updated estimateMemorySize() implementation to not include the size of VertexDescription. Once the PR is approved, what's the process of getting the changes included in a release?

randallwhitman commented:
Wow, that was really quick response to all our feedback in initial code review - Thanks @mbasmanova !

I'll defer to Sergey on approval and merge.

stolstov commented:
Everything looks good to me, just a small correction maybe for the OGCGeometry to exclude the spatial reference from the size calculation.

stolstov commented:
@mbasmanova When do you need this in a release?

mbasmanova commented:
@stolstov I updated the PR to exclude the size of the SpatialReference object from the estimate and added a comment. 

> When do you need this in a release?

Sergey, I'd like to use the new API in https://github.com/prestodb/presto/pull/9474 . Any chance a release can happen in the next couple of weeks? What's the usual cadence of the releases?

stolstov commented:
@mbasmanova Our releases happen as needed for the [Spatial Framework for Hadoop](https://github.com/Esri/spatial-framework-for-hadoop). The releases are done by @randallwhitman. This is a minor change, maybe Randall could make a release for you.

stolstov commented on src/main/java/com/esri/core/geometry/SizeOf.java:
```diff
@@ -0,0 +1,104 @@
+package com.esri.core.geometry;
```
@mbasmanova Could you add a copyright header here? See here Copyright and Licensing https://github.com/esri/contributing.

mbasmanova commented:
@stolstov Added copyright header to SizeOf.java.

stolstov commented:
@mbasmanova Maria, thank you for the contribution.

mbasmanova commented:
@stolstov Sergey, thanks for quick turnaround. I'm looking forward to use the new API in Presto.

stolstov commented on pom.xml:
```diff
@@ -100,6 +100,7 @@
 		<!-- dependency versions -->
 		<jackson.version>2.6.5</jackson.version>
 		<junit.version>4.12</junit.version>
+		<jol.version>0.2</jol.version>
```
@mbasmanova Is there a particular reason you used version 0.2? It seems the latest is 0.9.

mbasmanova replied:
@stolstov Sergey, no particular reason. I used the same version as Presto does.
